src/tools/rbd_mirror/image_map/Policy.h

```diff
@@ -0,0 +1,185 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_POLICY_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_POLICY_H
+
+#include <map>
+#include <tuple>
+
+#include "common/RWLock.h"
+#include "cls/rbd/cls_rbd_types.h"
+#include "include/rados/librados.hpp"
+
+#include "StateTransition.h"
+
+class Context;
+
+namespace librbd { class ImageCtx; }
+
+namespace rbd {
+namespace mirror {
+
+namespace image_map {
+
+class Policy {
+public:
+  Policy(librados::IoCtx &ioctx);
+
+  virtual ~Policy() {
+  }
+
+  struct Action {
+  public:
+    static Action create_add_action(Context *on_map, Context *on_finish);
+    static Action create_remove_action(Context *on_unmap, Context *on_purge,
+                                       Context *on_finish);
+    static Action create_shuffle_action(Context *on_map, Context *on_unmap,
+                                        Context *on_finish);
+
+    StateTransition::ActionType get_action_type() const;
+
+    void set_current_sub_action(StateTransition::SubActionType sub_action_type);
+    StateTransition::SubActionType get_current_sub_action() const;
+
+    void reset_action();
+
+    void execute_state_callback();
+    void execute_completion_callback(int r);
+
+    inline bool operator==(const Action &rhs) const {
+      return action_type == rhs.get_action_type() &&
+        current_sub_action == rhs.get_current_sub_action();
+    }
+    inline bool operator!=(const Action &rhs) const {
+      return !(*this == rhs);
+    }
+
+  private:
+    Action(StateTransition::ActionType action_type);
+
+    StateTransition::ActionType action_type;                         // action type for this action
+    StateTransition::SubActionType current_sub_action;               // currently executing subaction
+    std::map<StateTransition::SubActionType, Context *> context_map; // map sub action type to context
+                                                                     // callback
+  };
+
+  // init -- called during initialization
+  void init(const std::map<std::string, cls::rbd::MirrorImageMap> &image_mapping);
+
+  // lookup an image from the map
+  std::string lookup(const std::string &global_image_id);
+
+  // shuffle images when instances are added/removed
+  void add_instances(const std::vector<std::string> &instance_ids,
+                     std::set<std::string> *remap_global_image_ids, bool shuffle);
```

Comment:
actually this can be removed, but not due to the reason of it always being `false` but because the map is empty when `Policy` is instantiated -- so `do_shuffle_add_instances()` would not do anything.